### PR TITLE
Resampling MVP

### LINKED ISSUE #1010: Resampling MVP
The aim is to provide an API and minimal feature set that can be extended to cover all of the (useful) functionality provided by [Pandas resample method](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.resample.html).

Long term we will need to support the functionality provided by the arguments `rule`, `closed`, `label`, `origin`, `offset`.

For non-trivial bucket boundaries (e.g. last Thursday of every month) we should leverage Pandas to generate the actual boundaries of interest to pass to the C++ layer. For simpler boundaries (e.g. minute bars) we can have a more compact representation, although this is not required for the MVP.

Proposed MVP:

- Do not support upsampling, only downsampling.
- Present the resampled data back to the user, with no option to write the data back to another symbol directly.
- Leverage Pandas to convert `rule`, `origin`, and `offset` into a list of pairs of UTC timestamps stored as `int64_t` nanoseconds since epoch representing the bucket boundaries.
- Pass the `closed` and `label` arguments directly through to the clause constructor.
- Use the `QueryBuilder` directly rather than adding syntactic-sugar methods to the `Library` or `NativeVersionStore` classes.
- Use "data driven" approach to empty buckets. i.e. only include buckets in the output for which there was an index value in the appropriate range.
- Have a single clause to handle resampling (as opposed to the 2-stage process for hash-based groupings) since the repartition would always be trivial for resampling.
- Static schema supported only
- No "named agg" equivalent, so only one aggregation possible per input column

e.g.
```
q = QueryBuilder()
q = q.resample(rule="T", closed="left", label="left", origin="epoch", offset=None).agg({"open": first, "close": "last"})
df = lib.read(sym, date_range=(t1, t2), query_builder=q).data
```

### PR DESCRIPTION
**Pull request by alexowens90**
Closes #1010

### CODE REVIEW

**Review by willdealtry — APPROVED** [review]

File: cpp/arcticdb/async/tasks.hpp [hunk 1]
```diff
@@ -405,12 +406,20 @@ struct MemSegmentProcessingTask : BaseTask {
...
-            entity_ids_ = clause->process(std::move(entity_ids_));
+        // TODO: Replace with commented out code once C++20 is reinstated
+        for (auto clause = clauses_.crbegin(); clause != clauses_.crend(); ++clause) {
+            entity_ids_ = (*clause)->process(std::move(entity_ids_));
 
-            if(clause->clause_info().requires_repartition_)
+            if((*clause)->clause_info().requires_repartition_)
                 break;
         }
+//        std::ranges::reverse_view reversed_clauses{clauses_};
```
[the remark this answers is not in the document]
Comment (alexowens90) [reply cpp/arcticdb/async/tasks.hpp:410]: Moved next to TODO to make clearer

File: cpp/arcticdb/pipeline/frame_slice.hpp [hunk 2]
```diff
@@ -188,6 +188,14 @@ struct RangesAndKey {
     RangesAndKey() = delete;
     ARCTICDB_MOVE_COPY_DEFAULT(RangesAndKey)
 
+    bool operator==(const RangesAndKey& right) const {
```
[the remark this answers is not in the document]
Comment (alexowens90) [reply cpp/arcticdb/pipeline/frame_slice.hpp]: Done

File: cpp/arcticdb/processing/clause.hpp [hunk 3]
```diff
@@ -409,6 +411,84 @@ struct AggregationClause {
...
+    void set_aggregations(const std::vector<NamedAggregator>& named_aggregators);
+
+    void set_date_range(timestamp date_range_start, timestamp date_range_end) {
+        date_range_.emplace(date_range_start, date_range_end);
+    }
+
+    bool index_range_outside_bucket_range(timestamp start_index, timestamp end_index) const;
+
+    // Advances the bucket boundary iterator to the end of the last bucket that includes a value from a row slice with the given last index value
+    void advance_bucket_past_value(const std::vector<timestamp>& bucket_boundaries,
```
[the remark this answers is not in the document]
Comment (alexowens90) [reply cpp/arcticdb/processing/clause.hpp]: Done

File: cpp/arcticdb/processing/clause.hpp [hunk 4]
```diff
@@ -409,6 +411,84 @@ struct AggregationClause {
...
+    ProcessingConfig processing_config_;
+    std::string rule_;
+    ResampleBoundary label_boundary_;
+    // This will contain the data range specified by the user (if any) intersected with the range of timestamps for the symbol
+    std::optional<TimestampRange> date_range_;
+    // Inject this as a callback in the ctor to avoid language-specific dependencies this low down in the codebase
+    std::function<std::vector<timestamp>(timestamp, timestamp, std::string_view, ResampleBoundary)> generate_bucket_boundaries_;
+    std::vector<timestamp> bucket_boundaries_;
+    std::vector<SortedAggregatorInterface> aggregators_;
+    std::string str_;
```
[the remark this answers is not in the document]
Comment (alexowens90) [reply cpp/arcticdb/processing/clause.hpp:427]: It isn't required for all clauses right now (e.g. some clauses are not user facing, such as `RemoveColumnPartitioning` and `ColumnStatsGeneration`)

File: cpp/arcticdb/processing/clause.hpp [hunk 5]
```diff
@@ -409,6 +411,84 @@ struct AggregationClause {
     [[nodiscard]] std::string to_string() const;
 };
 
+template<ResampleBoundary closed_boundary>
+struct ResampleClause {
+    ClauseInfo clause_info_;
+    std::shared_ptr<ComponentManager> component_manager_;
+    ProcessingConfig processing_config_;
+    std::string rule_;
```
[the remark this answers is not in the document]
Comment (alexowens90) [reply cpp/arcticdb/processing/clause.hpp:419]: We support multiples/combinations of frequency strings such as `1h30mins` so an enum will not work

File: cpp/arcticdb/processing/processing_unit.cpp [hunk 6]
```diff
@@ -100,4 +100,55 @@ VariantData ProcessingUnit::get(const VariantNode &name) {
...
+            }
+        }
+    }
+    std::vector<ProcessingUnit> output;
+    output.reserve(output_map.size());
+    for (auto&& [_, processing_unit]: output_map) {
+        output.emplace_back(std::move(processing_unit));
+    }
+    // The expected get counts for all segments in a row slice should be the same
+    // This should always be 1 or 2 for the first row slice, and 1 for all of the others
```
[the remark this answers is not in the document]
Comment (alexowens90) [reply cpp/arcticdb/processing/processing_unit.cpp]: Changed to an assertion

**Pull request merged**